### PR TITLE
Add USE_CUSTOM_PG_BUILD option

### PR DESCRIPTION
**Pull request by byucesoy**
With this commit, we add USE_CUSTOM_PG_BUILD option to be able to run tests
against custom compiled PostgreSQL build. If this environment variable evaluates
to false or it is undefined, latest available PostgreSQL release will be
installed from Debian repositories.

### CODE REVIEW

File: travis/install_pg [hunk 1]
```diff
@@ -4,13 +4,31 @@
...
+    packages="$packages postgresql-$PGVERSION postgresql-server-dev-$PGVERSION"
+  fi
+
+  sudo apt-get -y -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" install $packages
+else
 
-sudo apt-get -y -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" install $packages
+  # clone PostgreSQL 
+  cd ~
+  git clone -b REL9_6_2 --depth 1 git://git.postgresql.org/git/postgresql.git
```
Comment (anarazel) [inline travis/install_pg]: It strikes me as a bad idea to hardcode versions here.  I think it'd be better to:
a) just use the relevant branch, instead of a tag
b) do this for both 9.5 and 9.6.

File: travis/install_pg [hunk 2]
```diff
@@ -4,13 +4,31 @@
...
+
+  # configure, build and install PostgreSQL
+  cd postgresql
+  ./configure --enable-cassert --with-openssl
```
Comment (anarazel) [inline travis/install_pg]: At least for valgrind this'll have to specify -DUSE_VALGRIND to enable full valgrind instrumentation. E.g. by doing export COPT='-DUSE_VALGRIND'.   I'm not sure whether we'd end up using the same infrastructure for that, but it's worth thinking about.

File: travis/install_pg [hunk 3]
```diff
@@ -4,13 +4,31 @@
...
+  make -j8 all
```
Comment (anarazel) [inline travis/install_pg]: Hardcoding concurrency might not be the greatest approach - I suspect few of the travis instances will have that many cores available.  Maybe use `-j$(grep -c ^processor /proc/cpuinfo)`?

File: travis/pg_travis_multi_test [hunk 4]
```diff
@@ -12,16 +12,31 @@ fi
...
-./configure --enable-coverage PG_CONFIG=/usr/lib/postgresql/$PGVERSION/bin/pg_config
+# If USE_CUSTOM_PG_BUILD variable is not set, we will assume custom build is
+# not required. If custom build is not required, that means PostgreSQL was
+# installed via apt, otherwise we compiled it from the source. Depending
+# on the situation, location of pg_config is different.
+if [ -z "$USE_CUSTOM_PG_BUILD" ] || ! [ $USE_CUSTOM_PG_BUILD ]; then
+  ./configure --enable-coverage PG_CONFIG=/usr/lib/postgresql/$PGVERSION/bin/pg_config
+else
+  ./configure --enable-coverage PG_CONFIG=/usr/local/pgsql/bin/pg_config
+fi
```
Comment (anarazel) [inline travis/pg_travis_multi_test]: Seems like it'd be better to set a $PG_CONFIG or such in the above branches, rather than repeating it later.  That way we don't need to maintain two branches here.

File: travis/pg_travis_multi_test [hunk 5]
```diff
@@ -12,16 +12,31 @@ fi
...
-# Run tests. DBs owned by non-standard owner put socket in /tmp
-make ${testtargets} || status=$?
+# If USE_CUSTOM_PG_BUILD variable is not set or set to false we will only run
+# testtargets, otherwise we will run check-vanilla and check isolation along
+# with our regular regression tests.
+if [ -z "$USE_CUSTOM_PG_BUILD" ] || ! [ $USE_CUSTOM_PG_BUILD ]; then
+  make ${testtargets} || status=$?
+else
+  make check && make check-vanilla && make check-isolation || status=$?
+fi
```
Comment (anarazel) [inline travis/pg_travis_multi_test]: Hm, is ignoring `$testtargets` a good idea?

File: travis/install_pg [hunk 6]
```diff
@@ -4,13 +4,31 @@
 
 set -eux
 
-# always install postgresql-common
-packages="postgresql-common libedit-dev libpam0g-dev libselinux1-dev"
+# If USE_CUSTOM_PG_BUILD variable is not set, we will assume custom build is
+# not required. If custom build is not required, we will install PostgreSQL
+# via apt, otherwise we will compile from the source.
+if [ -z "$USE_CUSTOM_PG_BUILD" ] || ! [ $USE_CUSTOM_PG_BUILD ]; then
```
Comment (anarazel) [inline travis/install_pg]: I'd suggest replacing this with `[ "$USE_CUSTOM_PG_BUILD" = "1" ]` or such.

Comment (byucesoy) [inline travis/install_pg]: I used `[ "${USE_CUSTOM_PG_BUILD-0}" = "0" ]`. Parameter substitution also handles if `USE_CUSTOM_PG_BUILD` is not set.

File: travis/install_pg (hunk 2 above)
Comment (byucesoy) [inline travis/install_pg]: We will probably never use valgrind on travis and adding -DUSE_VALGRIND also requires installing valgrind package for compilation. I do not want to install a package with no benefit.

If we want to run valgrind in the future we can add it to these scripts easily.

File: travis/install_pg [hunk 7]
```diff
@@ -4,13 +4,38 @@
 
 set -eux
 
-# always install postgresql-common
-packages="postgresql-common libedit-dev libpam0g-dev libselinux1-dev"
+# If USE_CUSTOM_PG_BUILD variable is not set, we will assume custom build is
+# not required. If custom build is not required, we will install PostgreSQL
+# via apt, otherwise we will compile from the source.
+if [ "${USE_CUSTOM_PG_BUILD-0}" = "0" ]; then
```
Comment (jasonmp85) [inline travis/install_pg]: Rather than adding conditionals and indenting everything, could you just add a new script (maybe `install_custom_pg`) and make each script exit early based on `USE_CUSTOM_PG_BUILD`'s value?

File: travis/install_pg [hunk 8]
```diff
@@ -4,13 +4,38 @@
...
+  fi
+
+  sudo apt-get -y -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" install $packages
+else
 
-sudo apt-get -y -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" install $packages
+  # clone PostgreSQL 
+  cd ~
+  if [ "$PGVERSION" = "9.5" ]; then
+    git clone -b REL9_5_STABLE --depth 1 git://git.postgresql.org/git/postgresql.git
```
Comment (jasonmp85) [inline travis/install_pg]: Just do `git clone -b REL${PGVERSION//./_}` to replace the dot in-line. no need for conditionals.

File: travis/install_pg [hunk 9]
```diff
@@ -4,13 +4,38 @@
...
+  elif [ "$PGVERSION" = "9.6" ]; then
+    git clone -b REL9_6_STABLE --depth 1 git://git.postgresql.org/git/postgresql.git
+  else
+    echo "Unsupported PostgreSQL version: $PGVERSION"
+    exit 128
+  fi
+
+  # configure, build and install PostgreSQL
+  cd postgresql
+  ./configure --enable-cassert --with-openssl
```
Comment (jasonmp85) [inline travis/install_pg]: These flags don't begin to actually emulate those used by a full PGDG build. @anarazel are we OK with that?

File: travis/install_pg [hunk 10]
```diff
@@ -4,13 +4,38 @@
...
+  make -j$(grep -c ^processor /proc/cpuinfo) -s all
```
Comment (jasonmp85) [inline travis/install_pg]: This argument to make should be set at the number of cores plus one:

```bash
procs="$(nproc)"
mjobs="$(expr $procs + 1)"
make -j${mjobs} # ...
```

File: travis/pg_travis_multi_test [hunk 11]
```diff
@@ -12,16 +12,33 @@ fi
...
 
-# Run tests. DBs owned by non-standard owner put socket in /tmp
-make ${testtargets} || status=$?
+# If USE_CUSTOM_PG_BUILD variable is not set or set to false we will only run
+# testtargets, otherwise we will run check-vanilla and check isolation along
+# with our regular regression tests.
+if [ "${USE_CUSTOM_PG_BUILD-0}" = "0" ]; then
+  make ${testtargets} || status=$?
+else
+  make ${testtargets} && make check-vanilla && make check-isolation || status=$?
```
Comment (jasonmp85) [inline travis/pg_travis_multi_test]: Don't call make three times here. In fact, it would be easiest to just check if `USER_CUSTOM_PG_BUILD` is set, then add `check-vanilla` and `check-isolation` to `testtargets`. Then you can use the existing invocation of `make` to do what you want.

File: travis/setup_apt [hunk 12]
```diff
@@ -4,18 +4,23 @@
 
 set -eux
 
-# import the PostgreSQL repository key
-sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys ACCC4CF8
+# If USE_CUSTOM_PG_BUILD variable is not set, we will assume custom build is
+# not required. If custom build is not required, we will setup apt to install
+# PostgreSQL via apt, otherwise we do not need to setup anything.
+if [ "${USE_CUSTOM_PG_BUILD-0}" = "0" ]; then
```
Comment (jasonmp85) [inline travis/setup_apt]: If `USER_CUSTOM_PG_BUILD` is set, just exit early.

**Comment by jasonmp85** [comment]
@anarazel — Would it make sense to simply _compile_ against PGDG (non-assert) packages, then compile and test against an assert build? If we run tests against an assert build, do we really need to run them against a non-assert build, as well?

I can see compiling against a non-assert build (since it might catch a needed assert-only macro), but am not clear on the benefits of testing against a non-assert build if we're already testing against assert builds.

File: travis/install_pg (hunk 9 above)
Comment (jasonmp85) [inline travis/install_pg]: @anarazel says to add `--enable-debug` and that's good for now.

**Comment by byucesoy** [comment]
@jasonmp85 I updated the PR based on your review.

File: travis/pg_travis_multi_test [hunk 13]
```diff
@@ -12,14 +12,30 @@ fi
 status=0
 testtargets="$*"
 
+# Depending on whether we compiled PostgreSQL from the source of not,
```
Comment (jasonmp85) [inline travis/pg_travis_multi_test]: I think you can get rid of this section.

File: travis/install_pg [hunk 14]
```diff
@@ -4,6 +4,13 @@
 
 set -eux
 
+# If USE_CUSTOM_PG_BUILD variable is set, we will compile PostgreSQL from the source.
+if [ "${USE_CUSTOM_PG_BUILD-0}" != "0" ]; then
+  install_custom_pg
```
Comment (jasonmp85) [inline travis/install_pg]: Rather than having one script call the other, can you just call _both_ in the `.travis.yml` file, and then have each have the "short-circuit" logic at the top to do the right thing?

This is how `install_citus` and `install_pg` work and we used to just call both.

File: travis/install_custom_pg [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+cd ~
+git clone -b REL${PGVERSION//./_}_STABLE --depth 1 git://git.postgresql.org/git/postgresql.git
+
+# we will use this to parallelize PostgreSQL compilation
+procs="$(nproc)"
+mjobs="$(expr $procs + 1)"
+
+# configure, build and install PostgreSQL
+cd postgresql
+./configure --enable-cassert --enable-debug --with-openssl
```
Comment (jasonmp85) [inline travis/install_custom_pg]: Add `--bindir="/usr/lib/postgresql/${PGVERSION}/bin"` to the `configure` arguments so you can just have the binaries live in the same spot whether PGDG or custom.

File: travis/install_pg [hunk 16]
```diff
@@ -4,6 +4,13 @@
 
 set -eux
 
+# If USE_CUSTOM_PG_BUILD variable is set, we will compile PostgreSQL from the source.
+if [ "${USE_CUSTOM_PG_BUILD-0}" != "0" ]; then
```
Comment (jasonmp85) [inline travis/install_pg]: For brevity, can you (a) abbreviate to `USE_CUSTOM_PG` and (b) just use the convention that if it's not defined or empty, it's false. Like this:

```bash
if [ -z "${USE_CUSTOM_PG} ]; then
  # ...
```

File: travis/setup_apt [hunk 17]
```diff
@@ -13,13 +19,6 @@ sudo rm /etc/apt/sources.list.d/google-chrome*
 # add the PostgreSQL 9.5 repository
 sudo sh -c 'echo "deb http://apt.postgresql.org/pub/repos/apt/ $(lsb_release -cs)-pgdg main 9.5" >> /etc/apt/sources.list.d/postgresql.list'
 
-if [ "${PGVERSION}" == "10" ]; then
```
Comment (jasonmp85) [inline travis/setup_apt]: Why was this removed?!?

File: travis/setup_apt [hunk 18]
```diff
@@ -4,6 +4,12 @@
 
 set -eux
 
+# If USE_CUSTOM_PG_BUILD variable is set, we will compile PostgreSQL
```
Comment (jasonmp85) [inline travis/setup_apt]: This is an optimization. I say don't modify this file at all…

**Comment by jasonmp85** [comment]
This is ready to go! Merging (and HLL is even working now).
